apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,874 @@
+        for (int i = 0; i < (int)all_vars.size(); i++) {
+            schedule_source << all_vars[i].name();
+            if (i != (int)all_vars.size() - 1) {
+                schedule_source << ",";
+            }
+        }
+        schedule_source << ")\n";
+        if (is_pure_def) {
+            if (gpu_thread_dim > 0) {
+                std::vector<Var> reordered_vars = vars;
+                std::swap(reordered_vars[0], reordered_vars[gpu_thread_dim]);
+                reorder_storage(func_or_stage, reordered_vars, schedule_source);
+            }
+        }
+    }
+    
+    if (!gpu_blocks.empty() || !r_gpu_blocks.empty()) { 
+        // Assign outer loops to GPU blocks
+        if (!fused_var.name().empty()) {
+            func_or_stage.gpu_blocks(fused_var);
+            schedule_source << "    .gpu_blocks(" << fused_var.name() << ")\n";
+        }
+        if (!fused_rvar.name().empty()) {
+            func_or_stage.atomic()
+                         .gpu_blocks(fused_rvar);
+            schedule_source << "    .atomic()\n";
+            schedule_source << "    .gpu_blocks(" << fused_rvar.name() << ")\n";
+        }
+        // Assign inner loops to GPU threads
+        if (!gpu_threads.name().empty()) {
+            func_or_stage.gpu_threads(gpu_threads);
+            schedule_source << "    .gpu_threads(" << gpu_threads.name() << ")\n";
+        }
+        if (!r_gpu_threads.name().empty()) {
+            func_or_stage.gpu_threads(r_gpu_threads);
+            schedule_source << "    .r_gpu_threads(" << r_gpu_threads.name() << ")\n";
+        }
+    } else {
+        // Not enough parallelism, use a single GPU thread
+        func_or_stage.gpu_single_thread();
+        schedule_source << "    .gpu_single_thread()\n";
+    }
+}
+
+
+template <typename FuncOrStage>
+void parallelize_vars_and_rvars_cpu(
+        const MachineParams &params,
+        FuncOrStage func_or_stage,
+        int natural_vector_size,
+        bool is_pure_def,
+        const std::vector<Var> &vars,
+        const std::vector<int> &var_bounds,
+        const std::vector<RVar> &rvars,
+        const std::vector<int> &rvar_bounds,
+        TailStrategy tail,
+        std::ostringstream &schedule_source) {
+    // Find the first variable that has bounds larger or equal than natural_vector_size,
+    // this is our vectorized dimension
+    const int split_size = natural_vector_size;
+    std::vector<Var> parallel_vars;
+    Var vectorized_var("");
+    int num_threads_var = 1;
+    int vectorized_dim = -1;
+    for (int i = 0; i < (int)vars.size(); i++) {
+        if (vectorized_var.name().empty() && var_bounds[i] >= split_size) {
+            vectorized_dim = i;
+            Var outer, inner;
+            func_or_stage.split(vars[i],
+                                outer,
+                                inner,
+                                split_size,
+                                tail);
+            schedule_source << "    .split(" <<
```

Comment:
Style elsewhere in the codebase is to put << at the start of each line, not the end

Reply:
Fixed.